### PR TITLE
Implementation of a Levenberg–Marquardt algorithm

### PR DESCRIPTION
**Pull request by CCsimon123**
Implementation of a Levenberg–Marquardt algorithm #119, according to the paper [Improvements to the Levenberg-Marquardt algorithm for nonlinear least-squares minimization](https://arxiv.org/abs/1201.5885).
The solver in `raphson.jl` was used as a template for this solver.

### CODE REVIEW

File: src/ad.jl [hunk 1]
```diff
@@ -26,7 +26,7 @@ end
 function SciMLBase.solve(prob::NonlinearProblem{<:Union{Number, StaticArraysCore.SVector},
                                                 iip,
                                                 <:Dual{T, V, P}},
-                         alg::Union{NewtonRaphson, TrustRegion},
+                         alg::Union{NewtonRaphson, TrustRegion, LevenbergMarquardt},
```
Comment (ChrisRackauckas) [inline src/ad.jl:29]:
```suggestion
                         alg::AbstractNewtonAlgorithm,
```

File: src/ad.jl [hunk 2]
```diff
@@ -35,7 +35,8 @@ end
 function SciMLBase.solve(prob::NonlinearProblem{<:Union{Number, StaticArraysCore.SVector},
                                                 iip,
                                                 <:AbstractArray{<:Dual{T, V, P}}},
-                         alg::Union{NewtonRaphson, TrustRegion}, args...;
+                         alg::Union{NewtonRaphson, TrustRegion, LevenbergMarquardt},
```
Comment (ChrisRackauckas) [inline src/ad.jl:38]:
```suggestion
                         alg::AbstractNewtonAlgorithm,
```

File: src/levenberg.jl [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+    @unpack u, p, λ, JᵀJ, DᵀD, J = cache
+
+    # Usual Levenberg-Marquardt step ("velocity").
+    cache.v = -(JᵀJ .+ λ .* DᵀD) \ mul!(cache.du_tmp, J', fu)
+
+    # Geodesic acceleration (step_size = v + a / 2).
+    @unpack v, alg = cache
+    h = alg.finite_diff_step_geodesic
+    f(cache.fu_tmp, u .+ h .* v, p)
+    cache.a = -J \ ((2 / h) .* ((cache.fu_tmp .- fu) ./ h .- mul!(cache.du_tmp, J, v)))
```
Comment (ChrisRackauckas) [inline src/levenberg.jl:306]: The two `\`'s should be changed to `linsolve` handlings.

File: src/levenberg.jl [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    if make_new_J
+        jacobian!(cache.J, cache)
+        mul!(cache.JᵀJ, cache.J', cache.J)
+        cache.DᵀD .= max.(cache.DᵀD, Diagonal(cache.JᵀJ))
+        cache.make_new_J = false
+    end
+    @unpack u, p, λ, JᵀJ, DᵀD, J = cache
+
+    # Usual Levenberg-Marquardt step ("velocity").
+    cache.v = -(JᵀJ .+ λ .* DᵀD) \ mul!(cache.du_tmp, J', fu)
```
Comment (ChrisRackauckas) [inline src/levenberg.jl:300]: `-(JᵀJ .+ λ .* DᵀD)` should be put into a temporary

Comment (ChrisRackauckas) [reply src/levenberg.jl:300]: `mul!(cache.du_tmp, J', fu)` I'd move that to a different line for clarity. Though it returns `cache.du_tmp` in most dispatches, it's usually best to mutate in one line and then use in the next

**Comment by CCsimon123** [comment]
Edit:
I will include this algorithm in `precompile_algs`. I naively thought I would have the same loading time as the 13 sec stated in #128, and when I got 20 sec for all three algorithms I thought the last 7 sec was due to the newly added `LevenbergMarquardt`.
And when I removed both `LevenbergMarquardt` and `TrustRegion` the loading time vanished.
But I now tested with only `LevenbergMarquardt` and `NewtonRaphson` and the loading time was good. While for `TrustRegion` and `NewtonRaphson` Is was 20 sec. Therefore I'll now include `LevenbergMarquardt` in the `precompile_algs`.

**Comment by CCsimon123** [comment]
Testing to remove the precompilation of `LevenbergMarquardt` to see if it fixes the error in the check "IntegrationTest / ModelingToolkit.jl/All/1 (pull_request)"

**Comment by CCsimon123** [comment]
Think I've found the issue from #128.
Have to go now, unfortunately, but I'll be back in 12h again.

File: src/levenberg.jl [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    # Geodesic acceleration (step_size = v + a / 2).
+    @unpack v, alg = cache
+    h = alg.finite_diff_step_geodesic
+    f(cache.fu_tmp, u .+ h .* v, p)
+
+    # The following lines do: cache.a = -J \ cache.fu_tmp
+    mul!(cache.du_tmp, J, v)
+    @. cache.fu_tmp = (2 / h) * ((cache.fu_tmp - fu) / h - cache.du_tmp)
+    linres = dolinsolve(alg.precs, linsolve, A = J, b = _vec(cache.fu_tmp),
+                        linu = _vec(cache.du_tmp), p = p, reltol = cache.abstol)
```
Comment (ChrisRackauckas) [inline src/levenberg.jl:320]: We may want to split the preconditioners between these two. But that's for follow up.

**Comment by ChrisRackauckas** [comment]
This looks good for a first pass. More to be done, but we should split correctness from improvements.

**Pull request merged**